File: Google Code Jam/2016/qround/coin_jam/coin_jam.cc
#include <cstdio>
#include <string>
#include <queue>
#include <cmath>
using namespace std;

int nonPrime(long long num) {
	for (int i = 2; i <= sqrt(num); i += 1) {
		//if (i % 2 == 0)
		//	i++;

		if (num % i == 0) {
			return i;
		}
	}
	return -1;
}

int main() {
	int T, N, J;
	int i, j, k;
	scanf("%d", &T);

	for (i = 1; i <= T; i++) {
		scanf("%d %d", &N, &J);
		printf("Case #%d:\n", i);

		int d[9] = {0};
		int found = 0;
		string s (N, '0');
		s[0] = '1';
		s[N-1] = '1';
		queue<string> q;
		q.push(s);
		while (found < J) {
			s = q.front(); q.pop();
			//printf("%s|", s.c_str());
			for (k = 1; k < s.length() && s[k] != '1'; k++) {
				string tmp = s;
				tmp[k] = '1';
				q.push(tmp);
			}
			for (k = 10; k >= 2; k--) {
				d[k - 2] = nonPrime(stoll(s, 0, k));
				if (d[k - 2] == -1)
					break;
			}
			if (k == 1) {
				printf("%s", s.c_str());
				for (k = 0; k < 9; k++) {
					printf(" %d", d[k]);
				}
				printf("\n");
				found++;
			}
		}
	}
}

File: Google Code Jam/2016/qround/coin_jam/coin_jam.py
from collections import deque
from math import sqrt

def nonPrime(num):
    i = 2
    while i < int(sqrt(num)) + 1:
        if num % i == 0:
            return i
        i += 1;
    return -1

def toNum(s, base):
    return reduce(lambda x,y: base * x + y, map(int, s))

def solve(n, j):
    s = ['1'] + ['0']*(n-2) + ['1'];
    q = deque([s])
    found = 0;
    while found < j:
        s = q.popleft()
        d = [None] * 9;
        for k in range(1, len(s)):
            if s[k] == '1':
                break
            tmp = s[:]
            tmp[k] = '1'
            q.append(tmp)
        for k in range(10, 1, -1):
            d[k - 2] = nonPrime(toNum(s, k))
            if d[k - 2] == -1:
                break
            if k == 2:
                k = k - 1
        if k == 1:
            found += 1
            print ''.join(s) + ' ' + ' '.join(map(str, d))

if __name__ == '__main__':
    t = int(raw_input().strip())
    for i in range(1, t+1):
        n, j = map(int, raw_input().strip().split())
        print 'Case #%d:' % i
        solve(n, j)